### PR TITLE
Feature/new tf model

### PR DESCRIPTION
**Pull request by mu-arkhipov**
TFModel refactored. NER, SlotFIll, and GoalOrientedBotNetwork rewritten to support new class.
What is done:
- TFModel refactored. Now it is more like regular tf model
- saving and restoring trainable variables (optimizer variables are excluded from saving and restoration)
- getting train_op become easy. batch_norm updates and gradients clipping are supported from the box 

Attention! To use load/save provided in TFModel the session must be attribute of the model with name `self.sess`!

### CODE REVIEW

File: deeppavlov/core/models/tf_model.py [hunk 1]
```diff
@@ -13,119 +13,123 @@
...
-        Define a single training step. Feed dict to tf session.
-        :param features: input features
-        :param args: any other inputs, including target vector, you need to pass for training
-        :return: metric to return, usually loss
         """
         pass
 
+    def load(self):
+        """Load model parameters from self.load_path"""
+        path = str(self.load_path.resolve())
```
Comment (yoptar) [inline deeppavlov/core/models/tf_model.py]: Why would you need to resolve save and load paths here? Are they not resolved in the Serializable class?

Comment (mu-arkhipov) [reply deeppavlov/core/models/tf_model.py]: save/load pathes are user_expanded but not resolved:
```def expand_path(path: Union[str, Path]) -> Path:
    return get_deeppavlov_root() / Path(path).expanduser()```

File: deeppavlov/core/models/tf_model.py [hunk 2]
```diff
@@ -13,119 +13,123 @@
...
+            # Exclude optimizer variables from saved variables
+            var_list = [var for var in tf.trainable_variables()
+                        if not var.name.startswith('Optimizer')]
+            saver = tf.train.Saver(var_list)
+            saver.restore(self.sess, path)
+
+    def save(self):
+        """Save model parameters to self.save_path"""
+        path = str(self.save_path.resolve())
+        print('[saving model to {}]'.format(path), file=sys.stderr)
```
Comment (yoptar) [inline deeppavlov/core/models/tf_model.py]: Use logging instead of prints, please.

File: deeppavlov/core/models/tf_model.py [hunk 3]
```diff
@@ -13,119 +13,123 @@
...
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
 
     @abstractmethod
-    def _add_placeholders(self):
-        """
-        Add all needed placeholders for a computational graph.
-        """
-        pass
+    def train_on_batch(self, x_batch, y_batch):
```
Comment (yoptar) [inline deeppavlov/core/models/tf_model.py]: This abstract method is defined in the parent class NNModel, so it's not needed here.

Comment (mu-arkhipov) [reply deeppavlov/core/models/tf_model.py]: The same thing with __call__, am I right?

Comment (yoptar) [reply deeppavlov/core/models/tf_model.py]: Yes

File: deeppavlov/models/ner/ner.py [hunk 4]
```diff
@@ -50,14 +44,11 @@ def __init__(self, **kwargs):
...
-        train_parameters = {par: opt[par] for par in train_parameters_names if par in opt}
+        train_parameters = {par: self.opt[par] for par in train_parameters_names if par in self.opt}
         self.train_parameters = train_parameters
 
-        self.opt = opt
-
         # Try to load the model (if there are some model files the model will be loaded from them)
-        super().__init__(save_path=save_path, load_path=load_path,
-                         train_now=train_now, mode=mode)
+        super().__init__(**self.opt)
```
Comment (yoptar) [inline deeppavlov/models/ner/ner.py]: Why do you send self.opt to super() and not kwargs? Also, in slotfill.py it's different

File: deeppavlov/core/models/tf_model.py [hunk 5]
```diff
@@ -13,119 +13,97 @@
...
-        """
-        return self._forward(instance, *args, **kwargs)
+    def load(self):
+        """Load model parameters from self.load_path"""
+        path = str(self.load_path.resolve())
+        # Check presence of the model files
+        if tf.train.checkpoint_exists(path):
+            log.info('[loading model from {}]'.format(path))
+            # Exclude optimizer variables from saved variables
+            var_list = [var for var in tf.trainable_variables()
```
Comment (vikmary) [inline deeppavlov/core/models/tf_model.py]: Maybe we should make variables' exclusion optional (via method parameter?)
Or
```python
def load(self, exclude_scopes=['Optimizer']):
    <...>
    var_list = self.get_trainable_variables(exclude_scopes)
    <..>
def _get_trainable_variables(scope_names=[]):
       all_vars = tf.trainable_variables()
       if scope_names:
            vars_to_train = set()
            for sn in scope_names:
                scope_vars = tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES, scope=sn)
                vars_to_train.update(scope_vars)
            return list(vars_to_train)
        else:
            return all_vars
```

Comment (mu-arkhipov) [reply deeppavlov/core/models/tf_model.py]:
```
def load(self, exclude_scopes=['Optimizer']):
    <...>
    var_list = self.get_trainable_variables(exclude_scopes)
    <..>
def _get_trainable_variables(exclude_scopes=[]):
       all_vars = tf.trainable_variables()
       vars_to_train = [var for var in all_vars if all(sc not in var.name for sc in exclude_scopes)]
       return vars_to_train
```

File: deeppavlov/core/models/tf_model.py [hunk 6]
```diff
@@ -13,119 +13,97 @@
...
         """
-        Load session from checkpoint
+        with tf.variable_scope('Optimizer'):
+            if learnable_scopes is None:
+                variables_to_train = tf.trainable_variables()
+            else:
+                variables_to_train = []
+                for scope_name in learnable_scopes:
+                    for var in tf.trainable_variables():
+                        if var.name.startswith(scope_name):
```
Comment (vikmary) [inline deeppavlov/core/models/tf_model.py]: `var.name.startswith(scope_name)` -> `scope_name not in var.name`?

File: deeppavlov/models/ner/network.py [hunk 7]
```diff
@@ -59,7 +54,8 @@ def __init__(self,
                  net_type='cnn',
                  char_filter_width=5,
                  verbouse=False,
```
Comment (vikmary) [inline deeppavlov/models/ner/network.py]: verbouse -> verbose?=)

File: deeppavlov/core/models/tf_model.py (hunk 6 above)
Comment (mu-arkhipov) [reply deeppavlov/core/models/tf_model.py]: done

File: deeppavlov/core/models/tf_model.py (hunk 5 above)
Comment (mu-arkhipov) [reply deeppavlov/core/models/tf_model.py]: in progress

File: deeppavlov/models/ner/network.py (hunk 7 above)
Comment (mu-arkhipov) [reply deeppavlov/models/ner/network.py]: fixed

File: deeppavlov/core/models/tf_model.py (hunk 5 above)
Comment (mu-arkhipov) [reply deeppavlov/core/models/tf_model.py]: Done
